File: search-a-2d-matrix/solution.swift
class Solution {
    func searchMatrix(_ matrix: [[Int]], _ target: Int) -> Bool {
        if matrix.isEmpty || matrix[0].isEmpty { return false }
        
        for row in matrix {
            if row[0] > target { continue }
            if bs(row, 0, row.count - 1, target) { return true }
        }
        
        return false
    }
    
    func bs(_ array: [Int], _ start: Int, _ end: Int, _ target: Int) -> Bool {
        guard start < end else { return array[start] == target }
        let middle = (start + end) / 2
        if array[middle] == target { return true }
        if array[middle] > target {
            return bs(array, 0, middle, target)
        } else {
            return bs(array, middle + 1, end, target)
        }
    }
}

File: search-a-2d-matrix/solution.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    if (matrix.length === 0 ) {
        return false; 
    }
    
    var len = matrix[0].length
    
    var bs = (row) => {
        return binarySearch(row, target, 0, len - 1)
    }
    
    for (var i in matrix) {
        if (matrix[i][len - 1] >= target) {
            return bs(matrix[i]); 
        } 
    }
    return false 
};

function binarySearch(nums, targ, start, end) {
    if (start > end ) {
        return false 
    }
    
    const mid = parseInt((start + end) / 2)
    
    var searchLeft = () => {
        return binarySearch(nums, targ, start, mid - 1)
    }
    
    var searchRight = () => {
        return binarySearch(nums, targ, mid + 1, end)
    }
    
    if (nums[mid] === targ) {
        return true; 
    } else if (nums[mid] > targ) {
        return searchLeft()
    } else if (nums[mid] < targ) {
        return searchRight()
    } else {
        return false 
    }
}
